tools/cli/NearObjectCli.cxx

```diff
@@ -237,63 +274,221 @@ NearObjectCli::AddSubcommandUwbMonitor(CLI::App* parent)
+    });
+
+    rawGetDeviceInfoApp->final_callback([this] {
+        auto uwbDevice = GetUwbDevice();
+        if (!uwbDevice) {
+            std::cerr << "no device found" << std::endl;
+            return;
+        }
+        if (!uwbDevice->Initialize()) {
+            std::cerr << "device not initialized" << std::endl;
+        }
+
+        m_cliHandler->HandleGetDeviceInfo(uwbDevice);
+    });
+
+    return rawGetDeviceInfoApp;
 }
 
 CLI::App*
 NearObjectCli::AddSubcommandUwbRangeStart(CLI::App* parent)
+{
+    auto& appConfigParamsData = m_cliData->appConfigParamsData;
+    auto rangeStartApp = parent->add_subcommand("start", "start ranging. Please refer to Table 29 of the FiRa UCI spec for more info on the options")->fallthrough();
+
+    // List mandatory params first
+    detail::AddEnumOption(rangeStartApp, appConfigParamsData.deviceRole, true);
+    detail::AddEnumOption(rangeStartApp, appConfigParamsData.multiNodeMode, true);
+    rangeStartApp->add_option("--NumberOfControlees", appConfigParamsData.numberOfControlees, "1 <= N <= 8")->capture_default_str()->required();
+
+    std::string deviceMacAddressDescription = "2-byte short MAC address of own device: e.g. 12:34";
+    std::string dstMacAddressDescription = "2-byte short MAC address(es) of other device(s). If device is Controller, list NumberOfControlees mac addresses. If device is Controlee, list Controller mac address";
+    if (appConfigParamsData.macAddressMode == uwb::UwbMacAddressType::Extended) { // TODO: Enable macAddressMode to be set before checking this value. Otherwise, the default empty value will always be used
+        deviceMacAddressDescription = "8-byte extended MAC address of own device: e.g. 12:34:56:78:87:65:43:21";
+        dstMacAddressDescription = "8-byte extended MAC address(es) of other device(s). If device is Controller, list NumberOfControlees mac addresses. If device is Controlee, list Controller mac address";
+    }
+    rangeStartApp->add_option("--DeviceMacAddress", m_cliData->deviceMacAddressString, deviceMacAddressDescription)->capture_default_str()->required();
+    rangeStartApp->add_option("--DestinationMacAddress", m_cliData->destinationMacAddressString, dstMacAddressDescription)->capture_default_str()->required();
+
+    detail::AddEnumOption(rangeStartApp, appConfigParamsData.deviceType, true);
+
+    rangeStartApp->parse_complete_callback([this, rangeStartApp] {
+        // TODO: Move validation logic into its own function
+
+        // Validate NumberOfControlees
+        if (m_cliData->appConfigParamsData.multiNodeMode == uwb::protocol::fira::MultiNodeMode::Unicast) {
+            if (m_cliData->appConfigParamsData.numberOfControlees != 1) {
+                std::cerr << "Only 1 controlee expected in Unicast mode" << std::endl;
+            }
+        } else {
+            if (m_cliData->appConfigParamsData.numberOfControlees < 1 || m_cliData->appConfigParamsData.numberOfControlees > 8) {
```

Comment:
Let's add a constant somewhere (probably `FiraDevice.hxx`) for max controlees so we can avoid magic numbers where possible, eg.

`constexpr auto MaxNumberOfControlees = 8`

Also, I might vaguely remember the spec says implementations may optionally support more than 8. If so, I wonder if we should not range check the inputs here and just pass the valeus onto the driver and let it do the checking?

Reply:
Hmm I don't remember that, at least not in the UCI spec. I'd like to keep the check here for now, but I'll definitely add the constant

Reply:
FYI, this is from CSML, table 7:
<img width="865" alt="image" src="https://user-images.githubusercontent.com/2082148/227624714-91adcc1f-43ec-4e10-ac76-8bd74f7b12e1.png">

Reply:
Oh I see, I can remove the check then